src/Npgsql/NpgsqlReadBuffer.Stream.cs

```diff
@@ -111,75 +111,94 @@ public override Task FlushAsync(CancellationToken cancellationToken)
 
             public override int Read(byte[] buffer, int offset, int count)
-                => Read(buffer, offset, count, CancellationToken.None, false).GetAwaiter().GetResult();
+                => Read(new Span<byte>(buffer, offset, count));
 
             public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
             {
                 using (NoSynchronizationContextScope.Enter())
-                    return Read(buffer, offset, count, cancellationToken, true).AsTask();
+                    return ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
             }
 
-            ValueTask<int> Read(byte[] buffer, int offset, int count, CancellationToken cancellationToken, bool async)
+#if !NET461 && !NETSTANDARD2_0
+            public override int Read(Span<byte> span)
+#else
+            public int Read(Span<byte> span)
+#endif
+            {
+                CheckDisposed();
+
+                var count = Math.Min(span.Length, _len - _read);
+
+                if (count == 0)
+                    return 0;
+
+                _buf.ReadBytes(span.Slice(0, count));
+                _read += count;
+
+                return count;
+            }
+
+#if !NET461 && !NETSTANDARD2_0
+            public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+#else
+            public ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+#endif
+
             {
                 CheckDisposed();
 
-                if (buffer == null)
-                    throw new ArgumentNullException(nameof(buffer));
-                if (offset < 0)
-                    throw new ArgumentNullException(nameof(offset));
-                if (count < 0)
-                    throw new ArgumentNullException(nameof(count));
-                if (buffer.Length - offset < count)
-                    throw new ArgumentException("Offset and length were out of bounds for the array or count is greater than the number of elements from index to the end of the source collection.");
                 if (cancellationToken.IsCancellationRequested)
                     return new ValueTask<int>(Task.FromCanceled<int>(cancellationToken));
 
-                count = Math.Min(count, _len - _read);
+                var count = Math.Min(buffer.Length, _len - _read);
 
                 if (count == 0)
                     return new ValueTask<int>(0);
 
-                var task = _buf.ReadBytes(buffer, offset, count, async);
-                if (task.IsCompleted) // This may be a bug in the new version of ValueTask
-                {
-                    _read += task.Result;
-                    return task;
-                }
-
-                return new ValueTask<int>(ReadLong(task, async));
+                var task = _buf.ReadBytes(buffer.Slice(0, count));
+                return ReadLong(task);
```

Comment:
I don't think this makes sense as it is... ReadLong can be folded into this method - please take a look.

Reply:
Yep, I improved that a little bit. Is it now OK?